### PR TITLE
Improve Round info on game card (fixes #3200)

### LINKED ISSUE #3200: 18MS: Incorrect round information in front page
Consider this game:

image

If I click on that game it turns out to be OR 7...
https://18xx.games/game/21238?action=187

So round information need to be corrected for 18MS.

### PR DESCRIPTION
**Pull request by perwestling**
This commit makes it possible for titles to control how the Round
information is presented on the game cards. To make this possible
we now starts to store round number as well as turn and round class.

The only title that uses this new method is 18MS.
It now shows the 18MS OR based on available turn and round num.
If round num has not yet been stored, it presents the possible
ORs as a range.
For Stock Round, no change.

Example:
  operating round, turn 4, round_num 2 => Operating 8
  stock round, turn 4 => Stock 4
  operating round, turn 4, round_num nil => Operating 7-8

### CODE REVIEW

File: assets/app/view/game/actionable.rb [hunk 1]
```diff
@@ -86,6 +86,7 @@ def process_action(action)
         if hotseat
           @game_data[:turn] = game.turn
           @game_data[:round] = game.round.name
+          @game_data[:round_num] = game.round.round_num
```
Comment (tobymao) [inline assets/app/view/game/actionable.rb:89]: what about for non hot seat??

File: assets/app/view/game_card.rb [hunk 2]
```diff
@@ -299,5 +299,14 @@ def render_broken
         ]),
       ])
     end
+
+    private
+
+    def game_card_round_info
+      game = Kernel.const_get(Engine::GAMES_BY_TITLE[@gdata['title']].to_s)
```
Comment (tobymao) [inline assets/app/view/game_card.rb:306]: isn't there a method in engine for this? why do you need to const_get

File: lib/engine/game/g_18_ms.rb [hunk 3]
```diff
@@ -200,10 +200,31 @@ def or_set_finished
...
       def or_description_short(turn, round)
+        G18MS.or_description_short(turn, round)
+      end
+
+      def self.or_description_short(turn, round)
         ((turn - 1) * 2 + round).to_s
       end
 
-      # Game will end directly after the end of OR 10
+      def self.game_card_round_info(gdata)
```
Comment (tobymao) [inline lib/engine/game/g_18_ms.rb:210]: game engines shouldn't take in gdata

File: assets/app/view/game/actionable.rb (hunk 1 above)
Comment (perwestling) [reply assets/app/view/game/actionable.rb:89]: Doesn't line 102 handle that? I looked where turn and round was stored, and only found these lines and line 101+103.

File: assets/app/view/game_card.rb (hunk 2 above)
Comment (perwestling) [reply assets/app/view/game_card.rb:306]: It creates a symbol, which I can use to call the class method with.

Example:
class Test
  def self.fum; end
end
a = "Test"
b = Kernel.const_get(a)

a.fum throws exception
b.fum works

Well, I did not find any prettier solution, but maybe there is?

File: lib/engine/game/g_18_ms.rb (hunk 3 above)
Comment (perwestling) [reply lib/engine/game/g_18_ms.rb:210]: Ok. I will pass the values on line 211-213 as parameters instead.

**Comment by tobymao** [comment]
i don't think we should do this approach at all. i think we should do a db migration to change what is stored in the game and have games set this string  directly from the engine

**Comment by perwestling** [comment]
Yes, that is probably better. The solution was kind of ugly.... So doing the migration sounds better.

**Comment by perwestling** [comment]
I put this as Draft but it can probably be closed.

**Comment by jenf** [comment]
Note pinned games won't update it properly

**Comment by perwestling** [comment]
New commit message:

**Improve Round info on game card (fixes #3200)**

Game keeps the current game info, and saves this the game is
stored.

The saved game info is used on game card. If game does not have
any it will show blank instead.

The round info is round name (minus Round) followed by the
or_description_short.
This means:
- 18MS show Operating 10 etc
- Standard 18xx show Operating 3.1 etc

The 18MS base class also contains a small refactoring which is
not connected to the fix.

File: lib/engine/game/base.rb [hunk 4]
```diff
@@ -75,7 +75,7 @@ def self.load(data, at_action: nil, actions: nil, pin: nil, optional_rules: nil,
     class Base
       attr_reader :raw_actions, :actions, :bank, :cert_limit, :cities, :companies, :corporations,
                   :depot, :finished, :graph, :hexes, :id, :loading, :loans, :log, :minors,
-                  :phase, :players, :operating_rounds, :round, :share_pool, :stock_market, :tile_groups,
+                  :phase, :players, :operating_rounds, :round, :round_info, :share_pool, :stock_market, :tile_groups,
```
Comment (tobymao) [inline lib/engine/game/base.rb:78]: why make this an attr? seems better as just a method that is computed dynamically

def round_info

**Comment by tobymao** [comment]
this doesn't migrate yet? also are you going to remove round and round_num since they may be unused?

**Comment by perwestling** [comment]
Here is how it looks like in game card, one game that I have done action in, and one that I have not:
image

**Comment by perwestling** [comment]
> this doesn't migrate yet? also are you going to remove round and round_num since they may be unused?

I run migration and it did not update round_info. Have not checked the details but is it so that game is not written down to database if unchanged? 
I will try to have round as a method instead and see if that makes a difference.

**Comment by tobymao** [comment]
the round should never be blank

**Comment by perwestling** [comment]
This is a new attempt to get this improved way of showing Round info.

> the round should never be blank

Don't we need to do this for broken games, or hotseat games? What should these display?

**Comment by perwestling** [comment]
Sorry, now I got it, leave out "Round: " as well if empty

**Comment by perwestling** [comment]
This is how it looks after latest change:

image

**Comment by perwestling** [comment]
Updated is left out if round_info is missing. 
So should I show Updated: always if not finished, or new?

**Comment by perwestling** [comment]
Pause this for now. Maybe pick it up at a later time.

**Pull request closed without merging**